include/fmt/ranges.h

```diff
@@ -0,0 +1,224 @@
+class is_like_std_string {
+	template<typename U> static auto check(U* p) -> decltype(
+		p->find('a')
+		, p->length()
+		, p->data()
+		, int());
+	template<typename  > static void check(...);
+public:
+	static const bool value = !std::is_void< decltype(check<T>(nullptr)) >::value;
+};
+
+template<typename T>
+constexpr bool is_like_std_string_v = is_like_std_string<T>::value;
+
+template<typename... Ts>
+struct conditional_helper {};
+
+template<typename T, typename _ = void>
+struct is_range_ : std::false_type {};
+
+template<typename T>
+struct is_range_<T,
+	std::conditional_t< false,
+	conditional_helper<
+	decltype(std::declval<T>().begin()),
+	decltype(std::declval<T>().end())
+	>, void>
+> : std::true_type {};
+
+template<typename T>
+constexpr bool is_range_v = is_range_<T>::value && !is_like_std_string<T>::value;
+
+
+/// tuple_size and tuple_element check.
+template<typename T>
+class is_tuple_like_ {
+	template<typename U> static auto check(U* p) -> decltype(
+		std::tuple_size< U >::value,
+		std::declval<typename std::tuple_element<0, U>::type>(),
+		int());
+	template<typename  > static void check(...);
+public:
+	static constexpr bool value = !std::is_void< decltype(check<T>(nullptr)) >::value;
+};
+
+template<typename T>
+constexpr bool is_tuple_like_v = is_tuple_like_<T>::value && !is_range_<T>::value;
+
+
+//=--------------------------------------------------------------------------------------------------------------------
+template<size_t... Is, class Tuple, class F>
+void for_each(std::index_sequence<Is...>, Tuple&& tup, F&& f) noexcept {
+	using std::get;
+	// using free function get<I>(T) now.
+	const int _[] = { 0,
+		((void)f(get<Is>(tup)),
+		0)... };
+	(void)_; // blocks warnings
+}
+//=--------------------------------------------------------------------------------------------------------------------
+template<class T>
+constexpr std::make_index_sequence<std::tuple_size<T>::value>
+get_indexes(T const&) { return {}; }
+
+//=--------------------------------------------------------------------------------------------------------------------
+template<class Tuple, class F>
+void for_each(Tuple&& tup, F&& f) {
+	const auto indexes = get_indexes(tup);
+	for_each(indexes, std::forward<Tuple>(tup), std::forward<F>(f));
+}
+
+} // namespace meta
+} // namespace fmt
+
+namespace fmt {
+
+// =====================================================================================================================
+template<typename TupleT, typename Char>
+struct formatter< TupleT, Char
+	, std::enable_if_t<fmt::meta::is_tuple_like_v<TupleT>> >
+{
+	fmt::formatting_tuple<Char> formating;
+
+	template <typename ParseContext>
+	FMT_CONSTEXPR auto parse(ParseContext &ctx) -> decltype(ctx.begin()) {
+		return formating.parse(ctx);
+	}
+
+	template <typename FormatContext = format_context>
+	auto format(const TupleT &values, FormatContext &ctx) -> decltype(ctx.out()) {
+		auto out = ctx.out();
+		std::ptrdiff_t i = 0;
```

Comment:
Why not size_t?

Reply:
Done, replaced by size_t.
I needed signed type in older implementation but now it can be size_t.